### PR TITLE
VM: Initial sources for the Hash Interpreter

### PR DESCRIPTION
**Pull request by feds01**
This PR adds the initial sources of the Hash Interpreter. Currently, the VM has a very primitive implementation of registers, stack and heap. It can also run very basic programs as demonstrated in the single test that exists.

The additions can be summarised as:
- Integrates the VM with `hash_pipeline` as a stage (this doesn't actually run anything in the VM, it just passes through as if it weren't present).
- Implemented a basic`Register` API
- Implemented a basic `Stack` API
- Implemented an Interpreter that can run a collection of instructions 
- Added a proof of concept test for the VM

This PR also fixes other unrelated issues:
- fix warning:`Deprecated AppSettings` issue with `clap`
- Add scaffolding for `hir` crate
- re-added `stack_size` option in command-line options

### CODE REVIEW

File: compiler/hash-vm/src/bytecode.rs [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    },
+    /// Logical AND (Like there exists a Non-Logical and)
+    And {
+        l1: Register,
+        l2: Register,
+    },
+    Not {
+        l1: Register,
+    },
+
```
Comment (kontheocharis) [inline compiler/hash-vm/src/bytecode.rs:205]: These should have 8,16,32,64 variants. Also they are bitwise, not logical.

File: compiler/hash-vm/src/bytecode.rs [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    },
+    Shl64 {
+        l1: Register,
+        l2: Register,
+    },
+    Shr64 {
+        l1: Register,
+        l2: Register,
+    },
+    /// Call a function at a given address?
```
Comment (kontheocharis) [inline compiler/hash-vm/src/bytecode.rs:248]: Remove question mark

File: compiler/hash-vm/src/bytecode.rs [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    Mov {
+        src: Register,
+        dest: Register,
+    },
+    Syscall {
+        id: Register,
+    },
+    Return,
+
+    /// The jump family (jumpin' around the christmas tree)
```
Comment (kontheocharis) [inline compiler/hash-vm/src/bytecode.rs:262]: rockin

File: compiler/hash-vm/src/bytecode.rs [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    },
+    JmpNeg {
+        l1: Register,
+        location: Register,
+    },
+    JmpZero {
+        l1: Register,
+        location: Register,
+    },
+    // Compare both values and store the result in `l1`
```
Comment (kontheocharis) [inline compiler/hash-vm/src/bytecode.rs:278]: Maybe document that this returns -1 | 1 | 0

File: compiler/hash-vm/src/error.rs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+pub type RuntimeResult<T> = Result<T, RuntimeError>;
+
+impl From<RuntimeError> for Report {
+    fn from(err: RuntimeError) -> Self {
+        match err {
+            RuntimeError::StackViolationAccess { kind, size, total } => {
+                Report {
+                    kind: ReportKind::Error,
+                    message: format!("Stack access violation occurred: tried to `{}` {}bytes from stack, but stack stack size is {}", kind, size, total ),
```
Comment (kontheocharis) [inline compiler/hash-vm/src/error.rs:37]: stack stack

File: compiler/hash-vm/src/register.rs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    pub fn set_register_b(&mut self, register: Register, value: &[u8; 1]) {
+        let reg = self.get_register_mut(register);
+        reg[7] = value[0];
+    }
+
+    /// Function to get a mutable reference to a  [Register] within the [RegisterSet].
+    fn get_register_mut(&mut self, register: Register) -> &mut [u8; 8] {
+        unsafe { self.registers.get_unchecked_mut(register.0 as usize) }
```
Comment (kontheocharis) [inline compiler/hash-vm/src/register.rs:102]: We ideally want no unsafes.

File: compiler/hash-vm/src/stack.rs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+            stack_pointer: 0,
+        }
+    }
+
+    /// Method that verifies that a particular call to modify the stack storage
+    /// is sane and safe.
+    pub fn verify_access(&self, access_kind: StackAccessKind, size: u8) -> RuntimeResult<()> {
+        match access_kind {
+            StackAccessKind::Pop if self.stack_pointer > size.into() => Ok(()),
+            StackAccessKind::Push if self.data.len() - self.stack_pointer > size.into() => Ok(()),
```
Comment (kontheocharis) [inline compiler/hash-vm/src/stack.rs:33]: Nice

**Review by kontheocharis — APPROVED** [review]

File: compiler/hash-vm/src/bytecode.rs (hunk 4 above)
Comment (feds01) [reply compiler/hash-vm/src/bytecode.rs:278]: This returns any range of values, the comparison should only jump if something is either zero, less than zero, or greater than zero.

**Review by kontheocharis — APPROVED** [review]

**Pull request merged**